[12.0] [ADD] Account Clearance Plan

qgroulard opened the pull request:
Adds the opportunity to create clearance plans in order to reorganize and spread debts (own or customer ones) over a longer period of time.

![clearance_plan](https://user-images.githubusercontent.com/43472442/80091640-bf897500-8561-11ea-9756-602bb6c8eb4d.png)

qgroulard committed:
[12.0] [ADD] Account Clearance Plan

sbidoul commented on account_clearance_plan/models/account_invoice.py:
```diff
@@ -0,0 +1,27 @@
+
+    def get_clearance_plan_wizard(self):
+        move_lines = self.env["account.move.line"]
+        for rec in self:
+            move_lines |= rec.move_id.line_ids.filtered(
+                lambda l: l.account_id == rec.account_id
+            )
```
nit:

```suggestion
        move_lines = self.mapped("move_id.line_ids").filtered(
            lambda l: l.account_id == rec.account_id
        )
```

sbidoul commented on account_clearance_plan/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,5 @@
+The module can also be used in a similar way for the user's company in case their suppliers offer them the opportunity to discuss a clearance plan.
+
+One can decide the debts to be part of the clearance plan (all partner's debts or some of them only) and split this amount into new financial commitments at new dates.
+Please notice that one clearance plan = one financial account.
```
```suggestion
Please notice that one clearance plan = one payable/receivable account.
```

sbidoul commented on account_clearance_plan/readme/USAGE.rst:
```diff
@@ -0,0 +1,11 @@
+To use this module, you need to:
+
+#. Go to *Invoicing > Accounting > Journal Items* (or any invoices)
```
```suggestion
#. Go to *Invoicing > Accounting > Journal Items*, or *Invoicing > Vendor Bills*, or *Invoicing > Customer Invoices*
```

sbidoul commented on account_clearance_plan/models/account_invoice.py:
```diff
@@ -0,0 +1,27 @@
+            move_lines |= rec.move_id.line_ids.filtered(
+                lambda l: l.account_id == rec.account_id
+            )
+        context = {"move_line_ids": move_lines.ids}
```
```suggestion
        context = {"active_ids": move_lines.ids, "active_model": "accout.move.line"}
```

Would this work and simplify the `default_get` in the wizard?

sbidoul commented on account_clearance_plan/wizard/account_clearance_plan.py:
```diff
@@ -0,0 +1,165 @@
+                    "move_id": move.id,
+                    key1: 0,
+                    key2: abs(line.amount_residual),
+                }
```
Don't we need a specific move line `name` here?

sbidoul commented on account_clearance_plan/wizard/account_clearance_plan.py:
```diff
@@ -0,0 +1,165 @@
+            new_lines |= line.with_context(check_move_validity=False).copy(
+                default={
+                    "move_id": move.id,
+                    key1: 0,
+                    key2: abs(line.amount_residual),
```
```suggestion
                    "debit": abs(line.amount_residual) if line.credit > 0 else 0,
                    "credit": abs(line.amount_residual) if line.debit > 0 else 0,
```

sbidoul commented on account_clearance_plan/wizard/account_clearance_plan.py:
```diff
@@ -0,0 +1,165 @@
+    )
+    clearance_plan_line_ids = fields.One2many(
+        comodel_name="account.clearance.plan.line", inverse_name="clearance_plan_id"
+    )
```
Do we need to populate `ref`, `narration` in the generated move?

qgroulard replied:
It's quite difficult to give a ref that makes sense, the clearance plan may be created based on many invoices or move lines. Do you have any suggestion ?

I've implemented the rest of your comments.

sbidoul replied:
Can the user provide it?

qgroulard replied:
Indeed !
The user can now provide it in the wizard as well as the `narration`.

sbidoul reviewed (commented):
Looks nice. A few minor comments.

We need to pay careful attention to the move and move line names that are generated so the partner ledger and credit control operations remain legible.

qgroulard committed:
[REF] Invoice Action

Action server has no property 'groups_ids', therefore refactored into an action window

gva-acsone commented:
Works really well ! 👍 

Just a small observation: it is currently impossible to open the 'clearance plan journal entry' starting from the invoice.
The 'view' button on the reconciled entry doesn't work. Would be really nice to make this work.

![Sans titre](https://user-images.githubusercontent.com/22000568/80613548-b9056c80-8a3d-11ea-9c0b-ca2a11559592.png)



For the rest everything seems to work just fine (functional test) and this functionality will be really useful.


Possible improvements for the future:

1. Add a field 'Amount still to allocate' above the Clearance Plan Board (difference between amount_to_allocate and amount_allocated)
2. Add a 'description' column in the clearance plan table. Each clearance plan line can be associated to a description, which will be automatically added in the 'label' field on the related journal item. This new field can be automatically pre-filled with the "Default Clearance Plan Move Line Name" defined in the accounting settings.

qgroulard committed:
[FIX] Invoice payment widget 'view' button

qgroulard committed:
[ADD] Improve usability

qgroulard commented:
> Works really well ! +1
> 
> Just a small observation: it is currently impossible to open the 'clearance plan journal entry' starting from the invoice.
> The 'view' button on the reconciled entry doesn't work. Would be really nice to make this work.
> 
> ![Sans titre](https://user-images.githubusercontent.com/22000568/80613548-b9056c80-8a3d-11ea-9c0b-ca2a11559592.png)
> 
> For the rest everything seems to work just fine (functional test) and this functionality will be really useful.
> 
> Possible improvements for the future:
> 
>     1. Add a field 'Amount still to allocate' above the Clearance Plan Board (difference between amount_to_allocate and amount_allocated)
> 
>     2. Add a 'description' column in the clearance plan table. Each clearance plan line can be associated to a description, which will be automatically added in the 'label' field on the related journal item. This new field can be automatically pre-filled with the "Default Clearance Plan Move Line Name" defined in the accounting settings.

@gva-acsone 
I've fixed the button and implemented your suggestions.
Thank you for the feedback !

qgroulard commented:
@OCA/core-maintainers , This PR is ready to merge

sbidoul commented:
/ocabot merge patch
